### PR TITLE
Add `unnecessary_to_owned` lint

### LINKED ISSUE #7933: Unnecessary call to `to_string` when the underlying type implements `AsRef<str>`
### What it does

Checks for expressions of the form `&X.to_string()`, used in a position requiring a `&str`, where `X` implements `AsRef<str>`.

In such a case, the call to `to_string` is unnecessary.

Similar to `to_string`/`AsRef<str>`, there's also `to_path_buf`/`AsRef<Path>` and `to_vec`/`AsRef<[T]>`.

There are probably others.

I am happy to tackle this if we can agree on the check, and whether it is useful.

### Categories (optional)

- Kind: perf

### Drawbacks

Potential "false positives" if `X`'s `AsRef<str>` and `to_string` implementations produce different strings. But I would expect such cases to be rare.

### Example

```rust
use std::{fmt::Write, io::Write as OtherWrite};

fn main() {
    let mut s = String::new();
    let dir = std::env::current_dir().unwrap();
    s.write_str(&dir.to_string_lossy().to_string()).unwrap();
    //                                ^^^^^^^^^^^^ unnecessary
    s.write_str(" is the current directory.\n").unwrap();
    std::io::stdout().write(s.as_bytes()).unwrap();
}
```

### PR DESCRIPTION
**Pull request by smoelius**
This PR adds a lint to check for unnecessary calls to `ToOwned::to_owned` and other similar functions (e.g., `Cow::into_owned`, `ToString::to_string`, etc.).

The lint checks for expressions of the form `&receiver.to_owned_like()` used in a position requiring type `&T` where one of the following is true:
* `receiver`'s type is `T` exactly
* `receiver`'s type implements `Deref<Target = T>`
* `receiver`'s type implements `AsRef<T>`

The lint additionally checks for expressions of the form `receiver.to_owned_like()` used as arguments of type `impl AsRef<T>`.

It would be nice if the lint could also check for expressions used as arguments to functions like the following:
```
fn foo<T: AsRef<str>>(x: T) { ... }
```
However, I couldn't figure out how to determine whether a function input type was instantiated from a parameter with a trait bound.

If someone could offer me some guidance, I would be happy to add such functionality.

Closes #7933

changelog: Add [`unnecessary_to_owned`] lint

### CODE REVIEW

**Comment by flip1995** [comment]
Oh this is a nice lint! I only skimmed the code. It'll probably take me a few days to do a in-depth review. In the mean time, could you add some documentation to the functions that makes it easier to understand what those functions are checking for, please? Especially the `check_call_arg` and the `check_addr_of_expr` functions.



> However, I couldn't figure out how to determine whether a function input type was instantiated from a parameter with a trait bound.

Shouldn't `implements_trait` on the arg `Ty` return `true` for the `AsRef` trait?

**Comment by smoelius** [comment]
> Oh this is a nice lint! ... In the mean time, could you add some documentation to the functions that makes it easier to understand what those functions are checking for, please?

Thank you. :) I will.

> Shouldn't `implements_trait` on the arg `Ty` return `true` for the `AsRef` trait? In other words: I don't think it should matter if you have `x: impl AsRef` or `<T: AsRef>(x; T)`. But maybe that's a bug/shortcoming of the `implements_trait` utility thinking

Sorry for not explaining well. Currently, I call [`fn_sig`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.fn_sig) to get the function's signature. For these two functions:
```rust
fn foo(_: impl AsRef<str>) {}
fn bar<T: AsRef<str>>(_: T) {}
```
`fn_sig` gives you structures with the following debug representations:
```rust
([impl AsRef<str>]; c_variadic: false)->()
([T]; c_variadic: false)->()
```
Currently, the lint catches calls to the `foo` function, because it parses the `ParamTy` name, which is `impl AsRef<str>` in the first debug representation above. But for the `bar` case, the `ParamTy` name is just `T`, so there's not enough information in the name to recognize that the argument has to implement `AsRef<str>`.

Presumably, there is some rustc API that allows one to query this, e.g., "What are the bounds on the type parameters of the function with detinition id `def_id`?" But I couldn't find that API.

**Comment by flip1995** [comment]
The [`FnSig::inputs`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/sty/struct.FnSig.html#method.inputs) method gives you a slice of `Ty`s. Calling `implements_trait(cx, arg_ty, as_ref_did, &[???])` should return `true` for `fn bar<T: AsRef<str>>(_: T) {}`? Or am I completely misunderstanding something?

**Comment by smoelius** [comment]
Well, it could be that I am missing something.

> The [`FnSig::inputs`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/sty/struct.FnSig.html#method.inputs) method gives you a slice of `Ty`s.

I think once you have `Ty`s it is too late. Because what you really want to know is: what were the requirements when those `Ty`s where chosen? Could we have chosen something else?

I'm not sure `implements_trait(.., as_ref_id, ..)` is of help here, because we can't be sure that `AsRef<..>` was a bound on the argument type.

That's sort of the crux of the issue: how do you find out what those bounds were?

**Comment by flip1995** [comment]
> I think once you have `Ty`s it is too late.

The `Ty` must retain all type information, so also all the traits this type implements. Once Clippy has access to `Ty`s, type resolution is fully done, so all relevant trait information for the type should be there.

So it doesn't really matter what bounds the argument type had in the code. After trait resolution and type checking, the generic type should implement the trait, no matter what actual type it is.

**Comment by smoelius** [comment]
Sorry, I think I may have just found the answer: https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.predicates_of

Please give me a few days to look into this and update the lint.

I apologize for my confusion.

**Comment by smoelius** [comment]
@flip1995 I think this is working as it should and you are welcome to start reviewing whenever you have time.

**Comment by camsteffen** [comment]
I just want to make sure we are careful with how this lint overlaps with `implicit_clone` and `redundant_clone`. This lint probably should share a list of methods with `implicit_clone`. I think this lint should just skip over cases of `redundant_clone`. Or you could make this lint pass responsible for those particular instances of `redundant_clone` if that simplifies the code overall.

**Comment by smoelius** [comment]
> I think this lint should just skip over cases of `redundant_clone`. Or you could make this lint pass responsible for those particular instances of `redundant_clone` if that simplifies the code overall.

How would you handle this, @camsteffen? Is there a way to, say, to call `redundant_clone`'s `check_fn` and check whether it fired?

**Comment by camsteffen** [comment]
Maybe you could do it the other way around - call the new lint implemention inside of redundant_clone.

**Comment by smoelius** [comment]
Actually maybe a better strategy is the following: have this lint apply only when the receiver is copy-able, or when the `to_owned`-like function is `Cow::into_owned`.

I think that would eliminate the possibility of overlap between `redundant_clone` and this lint, but would otherwise handle all cases that this lint currently tests for.

An alternative might be to do just the first part and extend `redundant_clone` to handle `Cow::into_owned`. But I'm not sure if that would go too much against the grain of `redundant_clone` (i.e., because in a call to `Cow::into_owned`, the receiver is moved).

**Comment by camsteffen** [comment]
I was just thinking of cases like `path_buf.to_path_buf()`.

**Comment by smoelius** [comment]
> I was just thinking of cases like `path_buf.to_path_buf()`.

OK. Then I will implement what I called a "better strategy" above. That should handle the case that you mentioned, and is less invasive than the other options (no changes to `redundant_clone`).

I will also implement the list of methods shared with `implicit_clone`.

@flip1995 Please give me a day or two to implement these changes.

**Comment by smoelius** [comment]
@flip1995 I think the just pushed changes address all of @camsteffen's concerns.

File: clippy_lints/src/unit_return_expecting_ord.rs [hunk 1]
```diff
@@ -169,7 +169,7 @@ impl<'tcx> LateLintPass<'tcx> for UnitReturnExpectingOrd {
                                     trait_name
                                 ),
                                 Some(last_semi),
-                                &"probably caused by this trailing semicolon".to_string(),
+                                "probably caused by this trailing semicolon",
```
Comment (hellow554) [inline clippy_lints/src/unit_return_expecting_ord.rs:172]: 🥳

File: tests/ui/unnecessary_to_owned.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    require_path(&path.to_path_buf());
+    require_path(&Cow::from(path).into_owned());
+    require_path(&path.to_owned());
+
+    require_str(&s.to_string());
+    require_str(&Cow::from(s).into_owned());
+    require_str(&s.to_owned());
+    require_str(&x.to_string());
+
+    require_slice(&slice.to_vec());
```
Comment (hellow554) [inline tests/ui/unnecessary_to_owned.rs:75]: What about the original `to_vec` problem: `for _ in slice.to_vec()`. Is this handled as well?

File: clippy_lints/src/methods/mod.rs [hunk 3]
```diff
@@ -1964,7 +1991,8 @@ impl_lint_pass!(Methods => [
     MANUAL_STR_REPEAT,
     EXTEND_WITH_DRAIN,
     MANUAL_SPLIT_ONCE,
-    NEEDLESS_SPLITN
+    NEEDLESS_SPLITN,
+    UNNECESSARY_TO_OWNED
```
[the remark this answers is not in the document]
Comment (smoelius) [reply clippy_lints/src/methods/mod.rs]: I agree.

**Comment by hellow554** [comment]
@smoelius can you please test the following code if the new lint triggers incorrectly?

```rust
#![feature(drain_filter)]

fn main() {
    let mut a = vec![3, 5, 7, -4, -6];
    for x in a.drain_filter(|x| *x < 0).collect::<Vec<_>>() {
        a.push(-x);
    }
}
```

The problem is, that I can't remove `collect`, because else I would get borrowchecker error.

**Comment by smoelius** [comment]
For me, the new lint doesn't seem to fire on that example. Could you please tell me what you are seeing?

**Comment by hellow554** [comment]
> For me, the new lint doesn't seem to fire on that example

that's great <3

> Please tell me if you observe something else.

I haven't observed anything, because I didn't want to compile clippy on my machine ;) That's why I asked you to run that code for me and as it turns out, it looks fine :)
Thanks!

**Review by llogiq — COMMENTED** [review]
I'm impressed by the very comprehensive test suite. :+1: I only found a few small nits. With those picked, I think this should be deemed mergeable.

File: clippy_lints/src/methods/mod.rs [hunk 4]
```diff
@@ -1885,6 +1887,32 @@ declare_clippy_lint! {
...
+    /// Checks for unnecessary calls to [`ToOwned::to_owned`](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html#tymethod.to_owned)
+    /// and other `to_owned`-like functions.
+    ///
+    /// ### Why is this bad?
+    /// The unnecessary calls result in unnecessary allocations.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let path = std::path::Path::new("x");
+    /// foo(&path.to_string_lossy().to_string());
```
[the remark this answers is not in the document]
Comment (smoelius) [reply clippy_lints/src/methods/mod.rs:1901]: `to_string` is handled outside of `is_clone_like`:
https://github.com/rust-lang/rust-clippy/blob/cb609a99044adc07ea4cdb0b4927f4ad461bb4b5/clippy_lints/src/methods/unnecessary_to_owned.rs#L390
https://github.com/rust-lang/rust-clippy/blob/cb609a99044adc07ea4cdb0b4927f4ad461bb4b5/clippy_lints/src/methods/unnecessary_to_owned.rs#L399-L401
The reason for this is that `implicit_clone` calls `is_clone_like` (as a consequence of this PR), and I didn't want to change `implicit_clone`'s behavior:
https://github.com/rust-lang/rust-clippy/blob/cb609a99044adc07ea4cdb0b4927f4ad461bb4b5/clippy_lints/src/methods/implicit_clone.rs#L15
Please let me know if you think something should be adjusted.

**Comment by smoelius** [comment]
@llogiq Thank for the review and for your kind words. Please let me know if you have any comments following the adjustments.

File: clippy_lints/src/methods/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (llogiq) [reply clippy_lints/src/methods/mod.rs:1901]: Then we should at least have a comment stating this.

[the remark this answers is not in the document]
Comment (smoelius) [reply clippy_lints/src/methods/mod.rs:1901]: Please tell me if what I wrote is inadequate.

**Comment by llogiq** [comment]
This looks good now. Thank you! @bors r+

**Pull request merged**